components/net_traits/image_cache.rs

```diff
@@ -224,272 +217,203 @@ impl CompletedLoad {
-                        let img_metadata = ImageMetadata { width: dimensions.width,
-                                                           height: dimensions.height };
-                        for listener in &pending_load.listeners {
-                            listener.respond(ImageResponse::MetadataLoaded(img_metadata.clone()));
-                        }
-                        pending_load.metadata = Some(img_metadata);
-                    }
-                }
-            }
-            (FetchResponseMsg::ProcessResponseEOF(result), key) => {
-                debug!("received EOF for {:?}", key);
-                match result {
-                    Ok(()) => {
-                        let pending_load = self.pending_loads.get_by_key_mut(&msg.key).unwrap();
-                        pending_load.result = Some(result);
-                        let bytes = pending_load.bytes.mark_complete();
-                        let sender = self.decoder_sender.clone();
-                        debug!("async decoding {} ({:?})", pending_load.url, key);
+#[derive(Copy, Clone, PartialEq, Hash, Eq, Deserialize, Serialize)]
+pub enum UsePlaceholder {
+    No,
+    Yes,
+}
 
-                        self.thread_pool.execute(move || {
-                            let msg = decode_bytes_sync(key, &*bytes);
-                            sender.send(msg).unwrap();
-                        });
-                    }
-                    Err(_) => {
-                        debug!("processing error for {:?}", key);
-                        match self.placeholder_image.clone() {
-                            Some(placeholder_image) => {
-                                self.complete_load(msg.key, LoadResult::PlaceholderLoaded(
-                                        placeholder_image))
-                            }
-                            None => self.complete_load(msg.key, LoadResult::None),
-                        }
-                    }
-                }
-            }
-        }
-    }
+// ======================================================================
+// ImageCache implementation.
+// ======================================================================
 
-    // Handle a message from one of the decoder worker threads
-    fn handle_decoder(&mut self, msg: DecoderMsg) {
-        let image = match msg.image {
-            None => LoadResult::None,
-            Some(image) => LoadResult::Loaded(image),
-        };
-        self.complete_load(msg.key, image);
-    }
+pub struct ImageCache {
+    // Images that are loading over network, or decoding.
+    pending_loads: Mutex<AllPendingLoads>,
+
+    // Images that have finished loading (successful or not)
+    completed_loads: Mutex<HashMap<ServoUrl, CompletedLoad>>,
+
+    // The placeholder image used when an image fails to load
+    placeholder_image: Option<Arc<Image>>,
 
+    // Webrender API instance.
+    webrender_api: Mutex<webrender_traits::RenderApi>,
```

Comment:
Multiple mutexes introduce opportunities for deadlocks. I would prefer to rely on Rust's borrow checker and write methods that accept &mut self arguments, rather than wrapping each member in a separate Mutex.